draft-ietf-ppm-dap.md

```diff
@@ -1554,12 +1563,18 @@ struct {
 where `prepare_resps` are the outbound prep steps computed in the previous step.
 The order MUST match `AggregationJobInitReq.prepare_inits`.
 
-The Helper responds to the Leader with HTTP status code 201 Created and a body
-consisting of the `AggregationJobResp`, with media type
-"application/dap-aggregation-job-resp".
+After receiving the response to its `AggregationJobInitReq`, the Leader makes
+requests to
+`GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` to check
+on the status of the aggregation job. If the aggregation job is ready, the
+Helper responds with HTTP status 200 OK and a body consisting of an
+`AggregationJobResp`, with media type "application/dap-aggregation-job-resp".
+If the aggregation job is not finished yet, the Helper responds with HTTP
+status 202 Accepted. The response MAY include a Retry-After header field to
+suggest a polling interval to the Leader.
 
 Changing an aggregation job's parameters is illegal, so further requests to
-`PUT /tasks/{tasks}/aggregation_jobs/{aggregation-job-id}` for the same
+`PUT /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for the same
 `aggregation-job-id` but with a different `AggregationJobInitReq` in the body
 MUST fail with an HTTP client error status code.
```

Comment:
This is very hard to implement for the helper. It requires possibly storing every body indefinitely. I think we can take advantage of the fact that aggregation is now async to make this better.

I propose that subsequent `PUT` requests to this enpoint carrying the same aggregation-job-id should be responded with the same response that previous requests were responded with regardless of , this means we have to, at most, store an (endpoint, response code) tuple. And we still maintain the idenpotency guarantees that we want, in case the leader retries the request

Reply:
I disagree here: in practice, the Helper only needs to store a hash of the previous request to perform request-equivalence checking. So the storage overhead isn't to store all request bodies indefinitely, it's to store the hash of the last request body until the next request is successfully processed.

I do think that checking that the request matches is very useful for "request/response coherency" reasons: if the Helper doesn't check the request body, we could end up in a situation where the Leader (due to a bug) repeats a request for some aggregation step with a different request, and the Helper would respond with the response body for the original request. This could be a very challenging situation to debug.

Reply:
> It requires possibly storing every body indefinitely.

Can you clarify the drawbacks to this? In Janus, [we accomplish this by storing a SHA-256 of the request body](https://github.com/divviup/janus/blob/main/db/00000000000001_initial_schema.up.sql#L225) which is fairly cheap in terms of storage, and I don't think has caused us problems (yet).

If we have to store `(endpoint, response code)` anyway, then we may as well go all the way and give more "truthful" idempotency (i.e. retain this clause), assuming that storing a request hash is not burdensome.

Reply:
You're right, this can be implemented with a hash. My follow up question is, what happens if the leader does repeat the request and keeps the same body? What should the helper do then? If this is specified I'm missing where

Reply:
I'm not seeing concrete mention of that scenario either. We've been treating repeated requests as idempotent, i.e. we return 201 Created.

We probably need language in this paragraph along the lines of "If the `AggregationJobInitReq` is identical to the original request, the Leader MAY respond with 201 Created". Thoughts from others?

Reply:
My view is that the discussion of PUT in [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110#section-9.3.4) makes it clear that such request are idempotent, but there's no downside to making that explicit here.

Reply:
@inahga let's mention this scenario explicitly.